### PR TITLE
Fix flake8 errors in codalab/

### PR DESCRIPTION
**Pull request by yiboliu**
### Reasons for making this change
Add fixing of errors in codalab/

### Related issues
#2660 

### Checklist

* [ ] I've added a screenshot of the changes, if this is a frontend change
* [ ] I've added and/or updated tests, if this is a backend change
* [x] I've run the [pre-commit.sh](https://github.com/codalab/codalab-worksheets/blob/master/pre-commit.sh) script
* [ ] I've updated [docs](https://github.com/codalab/codalab-worksheets/tree/master/docs), if needed

### CODE REVIEW

File: codalab/bundles/uploaded_bundle.py [hunk 1]
```diff
@@ -26,8 +26,9 @@ class UploadedBundle(NamedBundle):
     )
     METADATA_SPECS.append(
         MetadataSpec('link_format', str, 'Link format of bundle. Can be equal to'
-            '"raw" or "zip" (only "raw" is supported as of now).', optional=True)
+                                         '"raw" or "zip" (only "raw" is supported as of now).', optional=True)
```
Comment (teetone) [inline codalab/bundles/uploaded_bundle.py:29]: Nit: Is this spacing too much? Maybe one less tab?

File: codalab/lib/bundle_util.py [hunk 2]
```diff
@@ -2,7 +2,10 @@
 import json
 
 from codalab.bundles import get_bundle_subclass
-from codalab.client.json_api_client import JsonApiClient, JsonApiRelationship
+from codalab.client.json_api_client import (
+    JsonApiClient,  # NOQA - JsonApiClient used in comments
```
Comment (teetone) [inline codalab/lib/bundle_util.py]: Can you clarify what this means?

File: codalab/worker/bundle_service_client.py [hunk 3]
```diff
@@ -97,12 +99,14 @@ def _authorize(self):
             'Content-Type': 'application/x-www-form-urlencoded',
             'X-Requested-With': 'XMLHttpRequest',
         }
-        request = urllib.request.Request(
+        the_request = urllib.request.Request(
```
Comment (teetone) [inline codalab/worker/bundle_service_client.py]: Can we rename this variable? Maybe `request_to_send`?

File: setup.cfg [hunk 4]
```diff
@@ -34,3 +34,22 @@ ignore_missing_imports = True
...
+# E731 - do not assign a lambda expression, use a def, doesn't apply to our code
+
+# F401 - module imported but not used
+
+# W503 - line break before binary operator, conflicting with black
+# W605 - invalid escape sequence, modifying will cause test failures
+
+ignore = E203,E712,E731,W503
+per-file-ignores =
+  #  JsonApiClient used in comments, black doesn't inline comments in that line
```
Comment (teetone) [inline setup.cfg]: Can you clarify what this means?

File: setup.cfg [hunk 5]
```diff
@@ -34,3 +34,22 @@ ignore_missing_imports = True
...
+  codalab/lib/bundle_util.py: F401,F841
+  #  Commented explicitly that we want to keep the imports.
+  codalab/server/rest_server.py: F401
```
Comment (teetone) [inline setup.cfg:51]: Can you update the comment to `# Keep the unused imports as they are used to route the rest service`?

File: setup.cfg [hunk 6]
```diff
@@ -34,3 +34,22 @@ ignore_missing_imports = True
 [flake8]
 max-line-length = 200
 exclude = venv/*,var/*
+
+# E203 - White space before ':', conflicting with black
+# E712 - comparison to False should be 'if cond is False:' or 'if not cond:', doesn't apply to our code
```
Comment (teetone) [inline setup.cfg]: Why are we ignoring `E712`? It seems important to have.

https://www.flake8rules.com/rules/E712.html

File: setup.cfg [hunk 7]
```diff
@@ -34,3 +34,22 @@ ignore_missing_imports = True
...
+# E731 - do not assign a lambda expression, use a def, doesn't apply to our code
+
+# F401 - module imported but not used
+
+# W503 - line break before binary operator, conflicting with black
```
Comment (teetone) [inline setup.cfg]: nit: update lines 44 and 38 with `(conflicts with black)`

File: setup.cfg [hunk 8]
```diff
@@ -34,3 +34,22 @@ ignore_missing_imports = True
 [flake8]
 max-line-length = 200
 exclude = venv/*,var/*
+
+# E203 - White space before ':', conflicting with black
+# E712 - comparison to False should be 'if cond is False:' or 'if not cond:', doesn't apply to our code
+# E731 - do not assign a lambda expression, use a def, doesn't apply to our code
```
Comment (teetone) [inline setup.cfg]: Can you omit `doesn't apply to our code`?

File: codalab/bundles/uploaded_bundle.py (hunk 1 above)
Comment (yiboliu) [reply codalab/bundles/uploaded_bundle.py:29]: oh, this indentation is suggested by pycharm, to be at the same place of the beginning of former part.

File: setup.cfg [hunk 9]
```diff
@@ -34,3 +34,22 @@ ignore_missing_imports = True
 [flake8]
 max-line-length = 200
 exclude = venv/*,var/*
+
```
Comment (teetone) [inline setup.cfg:37]: Thanks for adding the error code comments. Can we organize the comments a bit better? Maybe something like this:

```
# Ignore completely:
# <flake8 code> - <description>
# <flake8 code> - <description>
#
# Ignore on a per file basis:
# <flake8 code> - <description>
# <flake8 code> - <description>
```

**Review by teetone — CHANGES_REQUESTED** [review]

File: codalab/lib/bundle_util.py (hunk 2 above)
Comment (yiboliu) [reply codalab/lib/bundle_util.py]: This means asking flake8 to ignore errors in this line. It's complaining about JsonApiClient is not used, but it's actually used in comments so I suppress it.

File: setup.cfg (hunk 4 above)
Comment (yiboliu) [reply setup.cfg]: The inline above comments was split into multiple lines then purpose of suppressing flake8 errors isn't served. So I do it here again. But would like to leave the inline comments there so other people can be aware of.

File: setup.cfg (hunk 6 above)
Comment (yiboliu) [reply setup.cfg]: This suggestion doesn't work for our code. Changing them could cause errors.

File: setup.cfg (hunk 4 above)
Comment (teetone) [reply setup.cfg]: Can you share the links to the lines that are problematic? I will take a look.

I think it's better to ignore a line or two within the file instead of ignoring a rule for the whole file.

File: setup.cfg (hunk 9 above)
Comment (yiboliu) [reply setup.cfg:37]: done

File: setup.cfg (hunk 8 above)
Comment (yiboliu) [reply setup.cfg]: done

File: setup.cfg (hunk 7 above)
Comment (yiboliu) [reply setup.cfg]: done

File: setup.cfg (hunk 5 above)
Comment (yiboliu) [reply setup.cfg:51]: done

File: codalab/lib/bundle_util.py (hunk 2 above)
Comment (teetone) [reply codalab/lib/bundle_util.py]: According to the flake8 [documentation](https://flake8.pycqa.org/en/3.1.1/user/ignoring-errors.html), should it be `# noqa: <error code>`?

File: codalab/worker/bundle_service_client.py (hunk 3 above)
Comment (yiboliu) [reply codalab/worker/bundle_service_client.py]: done

File: setup.cfg (hunk 4 above)
Comment (yiboliu) [reply setup.cfg]: The module is imported in this [line](https://github.com/codalab/codalab-worksheets/blob/master/codalab/lib/bundle_util.py#L5), and it's used [here](https://github.com/codalab/codalab-worksheets/blob/master/codalab/lib/bundle_util.py#L56).

Comment (yiboliu) [reply setup.cfg]: The problem is if I add the inline comments to ignore it, then the black will split my comments to the following lines and that will make it won't work.

File: codalab/lib/bundle_util.py (hunk 2 above)
Comment (yiboliu) [reply codalab/lib/bundle_util.py]: Another thing I can try would be to just leave the comment as `# noqa: F401` but can't add the description/explanation on the same line, have to do it in another line. How does that sound?

Comment (teetone) [reply codalab/lib/bundle_util.py]: Yeah I think you can just add a comment before.

File: setup.cfg (hunk 4 above)
Comment (teetone) [reply setup.cfg]: Yeah we shouldn't ignore a rule for an entire file for just that one line.

File: setup.cfg (hunk 6 above)
Comment (teetone) [reply setup.cfg]: I think that's a problem with our code then and it should be fixed.

Comment (yiboliu) [reply setup.cfg]:
```
>>> None is False
False
>>> None == False
False
>>> not None
True
>>> 0 == False
True
>>> 0 is False
False
```
This is the case that the suggestion doesn't work well.

File: codalab/lib/crypt_util.py [hunk 10]
```diff
@@ -186,8 +186,8 @@ def pbkdf2(password, salt, iterations, dklen=0, digest=None):
             dklen = hlen
         if dklen > (2 ** 32 - 1) * hlen:
             raise OverflowError('dklen too big')
-        l = -(-dklen // hlen)
-        r = dklen - (l - 1) * hlen
+        remainder = -(-dklen // hlen)
```
Comment (teetone) [inline codalab/lib/crypt_util.py]: `//` is floor division, so it doesn't return the remainder.

File: codalab/lib/worksheet_util.py [hunk 11]
```diff
@@ -64,11 +64,15 @@
 WORKSHEET_ITEM_TYPES = (TYPE_MARKUP, TYPE_DIRECTIVE, TYPE_BUNDLE, TYPE_WORKSHEET)
 
 
-BUNDLE_REGEX = re.compile('^\s*(\[(.*)\])?\s*\{([^{]*)\}\s*$')
-SUBWORKSHEET_REGEX = re.compile('^\s*(\[(.*)\])?\s*\{\{(.*)\}\}\s*$')
+BUNDLE_REGEX = re.compile(
+    '^\s*(\\[(.*)\\])?\s*\\{([^{]*)\\}\s*$'  # NOQA - modifying escapes could create end-to-end test failure
```
Comment (teetone) [inline codalab/lib/worksheet_util.py]: I think this is wrong formatting according to flake8. See my comment [here](https://github.com/codalab/codalab-worksheets/pull/2844#discussion_r495536795).

File: codalab/model/bundle_model.py [hunk 12]
```diff
@@ -848,7 +848,10 @@ def transition_bundle_starting(self, bundle, user_id, worker_id):
             # Check if the designated worker is going to be terminated soon
             row = connection.execute(
                 cl_worker.select().where(
-                    and_(cl_worker.c.worker_id == worker_id, cl_worker.c.is_terminating == False)
+                    and_(
+                        cl_worker.c.worker_id == worker_id,
+                        cl_worker.c.is_terminating == False,  # NOQA
```
Comment (teetone) [inline codalab/model/bundle_model.py]: Same with this one. There is no error code and comment explaining why we are ignoring the rule.

File: codalab/rest/interpret.py [hunk 13]
```diff
@@ -68,7 +62,7 @@ def _interpret_search():
     }
     ```
     """
-    return interpret_search(request.json)
```
Comment (teetone) [inline codalab/rest/interpret.py]: Why is this being modified?

File: codalab/worker/docker_utils.py [hunk 14]
```diff
@@ -26,13 +26,13 @@
 
 # This specific error happens when a user specifies an image with an incompatible version of CUDA
 NVIDIA_MOUNT_ERROR_REGEX = (
-    '[\s\S]*OCI runtime create failed[\s\S]*stderr:[\s\S]*nvidia-container-cli: '
-    'mount error: file creation failed:[\s\S]*nvidia-smi'
-)
+    '[\s\S]*OCI runtime create failed[\s\S]*stderr:[\s\S]*nvidia-container-cli: '  # NOQA
+    'mount error: file creation failed:[\s\S]*nvidia-smi'  # NOQA
+)  # Modifying escapes here could cause unit test failure.
```
Comment (teetone) [inline codalab/worker/docker_utils.py]: Is this causing failures in `docker_utils_test.py`? I think we should fix this. It seems a bit odd that we are modifying  `MEMORY_LIMIT_ERROR_REGEX`, but not `NVIDIA_MOUNT_ERROR_REGEX`.

File: setup.cfg [hunk 15]
```diff
@@ -34,7 +34,18 @@ ignore_missing_imports = True
...
+# Ignore completely:
+# E203 - White space before ':', (conflicts with black)
+# E731 - do not assign a lambda expression, use a def
+#
```
Comment (teetone) [inline setup.cfg:41]: nit: can we get rid of this extra line here and on line 43?

File: codalab/rest/interpret.py (hunk 13 above)
Comment (yiboliu) [reply codalab/rest/interpret.py]: I can't find any method with name interpret_search, and both PyCharm and Flake8 are complaining.

File: codalab/worker/docker_utils.py (hunk 14 above)
Comment (yiboliu) [reply codalab/worker/docker_utils.py]: Not really. It's always the end-to-end test is failing. I checked all the log files in the failing PR. The only message I got is `400 Client Error: Bad Request ("Unknown runtime specified nvidia")`, which matches none of these regexes.

Comment (teetone) [reply codalab/worker/docker_utils.py]: Which unit test is failing?

Comment (yiboliu) [reply codalab/worker/docker_utils.py]: Oh it's not unit test. It's end-to-end(frontend) test. Like what I ran here: https://github.com/codalab/codalab-worksheets/actions/runs/281351629

File: codalab/lib/worksheet_util.py [hunk 16]
```diff
@@ -547,7 +547,7 @@ def apply_func(func, arg):
                 t = tokens[2].replace(esc_slash, '/')
                 arg = re.sub(s, t, arg)
             elif f.startswith('['):  # substring
-                m = re.match('\[(.*):(.*)\]', f)
+                m = re.match('\\[(.*):(.*)\\]', f)
```
Comment (teetone) [inline codalab/lib/worksheet_util.py]: Can you undo this change for now?

**Review by teetone — APPROVED** [review]
